src/main/java/org/opensearch/neuralsearch/query/AgenticSearchQueryBuilder.java

```diff
@@ -0,0 +1,165 @@
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(this.queryText);
+        out.writeOptionalStringCollection(this.queryFields);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder xContentBuilder, Params params) throws IOException {
+        xContentBuilder.startObject(NAME);
+        if (Objects.nonNull(QUERY_TEXT_FIELD)) {
+            xContentBuilder.field(QUERY_TEXT_FIELD.getPreferredName(), queryText);
+        }
+        if (Objects.nonNull(queryFields) && !queryFields.isEmpty()) {
+            xContentBuilder.field(QUERY_FIELDS.getPreferredName(), queryFields);
+        }
+        xContentBuilder.endObject();
+    }
+
+    /**
+     * Creates AgenticSearchQueryBuilder from XContent.
+     * The expected parsing form looks like:
+     * {
+     *  "agentic": {
+     *    "query_text": "string",
+     *    "query_fields": ["string", "string"..]
+     *    }
+     * }
+     *
+     */
+
+    public static AgenticSearchQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        AgenticSearchQueryBuilder agenticSearchQueryBuilder = new AgenticSearchQueryBuilder();
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (QUERY_TEXT_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    agenticSearchQueryBuilder.queryText = parser.text();
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "Unknown field [" + currentFieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (QUERY_FIELDS.match(currentFieldName, parser.getDeprecationHandler())) {
+                    List<String> fieldsList = new ArrayList<>();
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                        fieldsList.add(parser.text());
+                    }
+                    agenticSearchQueryBuilder.queryFields = fieldsList;
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "Unknown field [" + currentFieldName + "]");
+                }
+            }
+        }
+
+        // Validate mandatory fields
+        if (agenticSearchQueryBuilder.queryText == null || agenticSearchQueryBuilder.queryText.trim().isEmpty()) {
+            throw new ParsingException(parser.getTokenLocation(), "[" + QUERY_TEXT_FIELD.getPreferredName() + "] is required");
+        }
+
+        return agenticSearchQueryBuilder;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        // No rewriting needed
```

Comment:
since no need to rewrite, we should throw exception here, so the query builder can save the effort travel through multiple doRewrite and reach to the doToQuery and finally throw exception. We knew the rewrite is handled by Processor then we should stop any doRewrite at this querybuilder

Reply:
doRewrite is called during query parsing and validation, returning this would do no change in the flow. We need exception thrown in `doToQuery`  because that's where lucene query is generated, if `doToQuery` is run that means the search pipeline didn't run. So throwing an exception there would makes sens